nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java

```diff
@@ -0,0 +1,238 @@
+            .Builder()
+            .name("icon-emoji")
+            .displayName("Icon Emoji")
+            .description("Icon Emoji to be used for the message. Must begin and end with a colon, e.g. :ghost:")
+            .required(false)
+            .expressionLanguageSupported(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .addValidator(new EmojiValidator())
+            .build();
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("FlowFiles are routed to success after being successfully sent to Slack")
+            .build();
+
+    public static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("FlowFiles are routed to failure if unable to be sent to Slack")
+            .build();
+
+    public static final List<PropertyDescriptor> descriptors = Collections.unmodifiableList(
+            Arrays.asList(WEBHOOK_URL, WEBHOOK_TEXT, CHANNEL, USERNAME, ICON_URL, ICON_EMOJI));
+
+    public static final Set<Relationship> relationships = Collections.unmodifiableSet(
+            new HashSet<>(Arrays.asList(REL_SUCCESS, REL_FAILURE)));
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    // Validate the channel (or username for a direct message)
+    private String validateChannel(String channel) {
+        if ((channel.startsWith("#") || channel.startsWith("@")) && channel.length() > 1) {
+            return null;
+        }
+        return "Channel must begin with '#' or '@'";
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if ( flowFile == null ) {
+            return;
+        }
+
+        try {
+            JsonObjectBuilder builder = Json.createObjectBuilder();
+            String text = context.getProperty(WEBHOOK_TEXT).evaluateAttributeExpressions(flowFile).getValue();
+            if (text != null && !text.isEmpty()) {
+                builder.add("text", text);
+            } else {
+                // Slack requires the 'text' attribute
+                throw new ProcessException("FlowFile should have non-empty " + WEBHOOK_TEXT.getName());
+            }
+
+            String channel = context.getProperty(CHANNEL).evaluateAttributeExpressions(flowFile).getValue();
+            if (channel != null && !channel.isEmpty()) {
+                String error = validateChannel(channel);
+                if (error == null) {
+                    builder.add("channel", channel);
+                } else {
+                    throw new ProcessException(String.format("Invalid channel '%s': %s", channel, error));
```

Comment:
Same comment as above, consider sending the flow file to failure rather than throwing an exception.